service/src/main.rs

```diff
@@ -687,46 +681,18 @@ fn print_events(events: Events, _sender: Sender<String>) {
 	}
 }
 
-pub fn init_parentchain_components<E: EnclaveBase + Sidechain>(
-	api: &ParentchainApi,
-	enclave_api: Arc<E>,
-) -> Result<Header, Error> {
-	let genesis_hash = api.get_genesis_hash().unwrap();
-	let genesis_header: Header = api.get_header(Some(genesis_hash)).unwrap().unwrap();
-	info!("Got genesis Header: \n {:?} \n", genesis_header);
-	if api.is_grandpa_available()? {
-		let grandpas = api.grandpa_authorities(Some(genesis_hash)).unwrap();
-		let grandpa_proof = api.grandpa_authorities_proof(Some(genesis_hash)).unwrap();
-
-		debug!("Grandpa Authority List: \n {:?} \n ", grandpas);
-
-		let authority_list = VersionedAuthorityList::from(grandpas);
-
-		let params = LightClientInitParams::Grandpa {
-			genesis_header,
-			authorities: authority_list.into(),
-			authority_proof: grandpa_proof,
-		};
-
-		Ok(enclave_api.init_parentchain_components(params).unwrap())
-	} else {
-		let params = LightClientInitParams::Parachain { genesis_header };
-
-		Ok(enclave_api.init_parentchain_components(params).unwrap())
-	}
-}
-
 /// Subscribe to the node API finalized heads stream and trigger a parent chain sync
 /// upon receiving a new header.
 fn subscribe_to_parentchain_new_headers<E: EnclaveBase + Sidechain>(
-	enclave_api: Arc<E>,
-	api: &ParentchainApi,
+	parentchain_handler: Arc<ParentchainHandler<ParentchainApi, E>>,
 	mut last_synced_header: Header,
 ) -> Result<(), Error> {
 	let (sender, receiver) = channel();
-	api.subscribe_finalized_heads(sender).map_err(Error::ApiClient)?;
+	parentchain_handler
+		.parentchain_api()
+		.subscribe_finalized_heads(sender)
+		.map_err(Error::ApiClient)?;
```

Comment:
Can't we add the `subscribe_finalized_heads` to the parentchain handler trait? Exposing members (like the api in this case) breaks encapsulation

Reply:
Sadly not.. Not without hardcoding the ParentchainApi. `subscribe_finalized_heads` is not available as a trait (https://github.com/scs/substrate-api-client/blob/ab586d8be719243aa29503ec8c7b1457b2cd0830/src/std/rpc/ws_client/mod.rs#L75-L97)

So I left it as it was for now.

Reply:
I'll add a TODO and create an issue on the api_client side

Reply:
Alright, that's good enough for me 👍

Reply:
There's actually an issue already: https://github.com/scs/substrate-api-client/issues/267 Linked it in the TODO comment as well.